### PR TITLE
Increase superuser reserved connections

### PR DESCRIPTION
**Pull request by l-wang**
As requested from field, 3 superuser connections is not enough for gpdb when customers run superuser maintenance scripts. 10 is the same value as the resource group admin_group's concurrency default limit. Also only have the fts connection reserved for primaries.

### CODE REVIEW

File: src/backend/utils/misc/guc.c [hunk 1]
```diff
@@ -1886,11 +1886,11 @@ static struct config_int ConfigureNamesInt[] =
 	{
 		{"superuser_reserved_connections", PGC_POSTMASTER, CONN_AUTH_SETTINGS,
 			gettext_noop("Sets the number of connection slots reserved for "
-						"superusers (including reserved FTS connections)."),
+						"superusers (including reserved FTS connection for primaries)."),
 			NULL
 		},
 		&ReservedBackends,
-		3, RESERVED_FTS_CONNECTIONS, MAX_BACKENDS,
+		10, RESERVED_FTS_CONNECTIONS, MAX_BACKENDS,
```
Comment (zhangh43) [inline src/backend/utils/misc/guc.c]: The comment of ReservedBackends in code is that
```
 * ReservedBackends is the number of backends reserved for superuser use.
 * This number is taken out of the pool size given by MaxBackends so
 * number of backend slots available to non-superusers is
 * (MaxBackends - ReservedBackends).  Note what this really means is
 * "if there are <= ReservedBackends connections available, only superusers
 * can make new connections" --- pre-existing superuser connections don't
 * count against the limit.
```
So it seems that super user could already make more than 3(10) connections if there are enough free connection slots.

**Comment by ashwinstar** [comment]
> On Feb 24, 2020, at 10:20 PM, Hubert Zhang <notifications@github.com> wrote:
> 
> ﻿
> @zhangh43 commented on this pull request.
> 
> In src/backend/utils/misc/guc.c:
> 
> >  			NULL
>  		},
>  		&ReservedBackends,
> -		3, RESERVED_FTS_CONNECTIONS, MAX_BACKENDS,
> +		10, RESERVED_FTS_CONNECTIONS, MAX_BACKENDS,
> The comment of ReservedBackends in code is that
> 
>  * ReservedBackends is the number of backends reserved for superuser use.
>  * This number is taken out of the pool size given by MaxBackends so
>  * number of backend slots available to non-superusers is
>  * (MaxBackends - ReservedBackends).  Note what this really means is
>  * "if there are <= ReservedBackends connections available, only superusers
>  * can make new connections" --- pre-existing superuser connections don't
>  * count against the limit.
> So it seems that super user could already make more than 3(10) connections if there are enough free connection slots.
> 
Problem is not when connections are available. It’s when non-superusers are using close to their limit and then reserved connections for superusers is not enough.

File: src/backend/utils/init/postinit.c [hunk 2]
```diff
@@ -599,13 +599,14 @@ BaseInit(void)
  */
 static void check_superuser_connection_limit()
 {
+	int limit = IS_QUERY_DISPATCHER()? 0 : RESERVED_FTS_CONNECTIONS;
 	if (!am_ftshandler &&
-		!HaveNFreeProcs(RESERVED_FTS_CONNECTIONS))
+		!HaveNFreeProcs(limit))
```
Comment (paul-guo-) [inline src/backend/utils/init/postinit.c]: !am_ftshandler && !S_QUERY_DISPATCHER() && HaveNFreeProcs(RESERVED_FTS_CONNECTIONS)  seems to be cleaner.

Comment (l-wang) [reply src/backend/utils/init/postinit.c]: OK updated.

**Comment by l-wang** [comment]
Merged the change via commit 8dab92cc38a929fab701db1e32677de9bf5a6903 and 38565eff10f466c2f79505fda8b34487b4138a34

**Pull request closed without merging**